dp/cloud/python/magma/configuration_controller/crl_validator/crl_validator.py

```diff
@@ -0,0 +1,162 @@
+
+        Returns: None
+        """
+        for host in self._certificates.keys():
+            certificate = get_certificate(hostname=host)
+            crls = get_certificate_crls(certificate=certificate)
+
+            self._certificates[host] = certificate
+            self._crls[certificate.serial_number] = crls
+
+
+class CRLValidator(object):
+    """
+    This class is responsible for validating given urls' SSL certs with their respective CRLs.
+    """
+
+    def __init__(
+            self, urls: List[str], certificates_update_rate: int = 300,
+    ) -> None:
+        """
+        Args:
+            urls: list of urls that we should prefetch certificates and CRLs for
+            certificates_update_rate: time in seconds between certificates and CRLs update
+        """
+        hosts = [get_host(url=url) for url in urls]
+
+        # Certificates and CRLs dicts are shared with separate updater thread.
+        # All we want is to have recent certs for given hosts, so usual thread related issues don't really bother us.
+        self._certificates = dict.fromkeys(hosts)
+        self._crls = {}
+
+        # Start updater thread to update certificates and CRLs in the background.
+        self._updater_thread = CertificatesUpdaterThread(
+            certificates=self._certificates,
+            crls=self._crls,
+            update_rate=certificates_update_rate,
+        )
+        self._updater_thread.start()
+
+    def is_valid(self, url: str) -> bool:
+        """ Check if given url's SSL certificate is not revoked by its Certificate Revocation Lists.
+
+        Args:
+            url: url string
+
+        Returns:
+            bool: False if certificate is not revoked
+
+        Raises:
+            SSLError: if certificate is revoked
+        """
+        host = get_host(url=url)
+        certificate = self._get_certificate(hostname=host)
+        crls = self._get_certificate_crls(certificate=certificate)
+        return not is_certificate_revoked(certificate=certificate, crls=crls)
+
+    def _get_certificate(self, hostname: str) -> x509.Certificate:
+        """ Get cached certificate for given host, fetch new if it does not exist.
+
+        Args:
+            hostname: host for which certificate was issued
+
+        Returns:
+            SSL certificate
+        """
+        if not self._certificates.get(hostname):
+            self._certificates[hostname] = get_certificate(hostname=hostname)
+        return self._certificates[hostname]
+
+    def _get_certificate_crls(
+            self, certificate: x509.Certificate,
+    ) -> List[x509.CertificateRevocationList]:
+        """ Get cached CertificateRevocationLists for given certificate, fetch new if they does not exist.
+
+        Args:
+            certificate: certificate that CRLs were attached to
+
+        Returns:
+            list of Certificate Revocation Lists
+        """
+        serial_number = certificate.serial_number
```

Comment:
possible race condition between updating certs and reading them

Reply:
I'm not sure it's that simple, because of two things:
1. Update is not doing atomic operations and also it is not guaranteed that even dict operations are thread safe (they are for example in cpython, but it is not guaranteed in any way), so for example you may be left in some inconsistent state (for example dict insert may require changing its internal structure and if you read something in the other thread it may not even the value you are looking for), which cal lead to weird undefined behavior. Also bug related to concurrency tend to be very difficult to debug, since they are not deterministic and also you have to consider multiple different scenarios of threads overlapping with each other
2. This function is doing not only reads, but also writes, so in fact if there is no cert for given serial number, then each call to this function would result in an additional query, which is what we were trying to avoid by introducing a background job

So I suggest a few things here:

1. Use mutex for `_crls` access/modification (or even better RWmutex, unfortunately not in python standard library), there are only up to 7 threads using it, so it should drop the performance that much, but I think it is very important to have thread safe code
2. Do not do any writes in this function, only reads. If we don't have cert for some serial number then, we don't get it, similarly (this is even the current behavior) if we have outdated cert we return it. This is the point of having background cert manager - not to query for new crls by ourselves.
4. This is more of a question - do you know how often CRLs needs to be updated? I've seen some guidelines saying that doing it once every 24h should be enough, so we don't have to worry about certs being outdated for 5 min and maybe we can even make refresh interval larger (for example 1h or more)

Reply:
1. Added `rwmutex.RWLock` for read and write to `_certificates` and `_crls`.
2. Removed the writes from `is_valid` method, if we don't have specific cert then `KeyError` is raised, informing that this specific host's certs are not available since this host was not given on validator initialization and it's certs were not prefetched. 
3. I've also seen the 24h mark. I think 1h cache on our end is perfectly reasonable, changed cache time in config to 1h.